File: lab3/spellchecker/src/Document.hs

{-# LANGUAGE DoAndIfThenElse #-}
-- 

module Document where

import System.IO
import Control.Monad
import Data.Char

type Word = String
data Document = Document Handle Handle

-- Abre los archivos especificados por los paths
-- pasados como argumentos. El primer path repre-
-- senta el archivo de entrada a procesar, y el
-- segundo path representa el archivo de salida
-- donde se guarda el documento ya procesado.
doc_open :: FilePath -> FilePath -> IO Document
doc_open fp1 fp2 = 
    do 
        file_in <- openFile fp1 ReadMode
        file_out <- openFile fp2 WriteMode
        return (Document file_in file_out)

-- Cierra los archivos especificados
doc_close :: Document -> IO ()
doc_close (Document f1 f2) = 
    do 
        hClose f1
        hClose f2
        return ()


constructWord :: Word -> Int -> Handle -> Handle -> IO Word  
constructWord word_to_return i file_in file_out =
    do 
        char_readed <- hGetChar file_in
        
        if isAlphaNum char_readed then
            constructWord (word_to_return++[char_readed]) i+1 file_in file_out
        else do
            if i == 0 then
                hPutChar file_out char_readed
            else
                hSeek file_in RelativeSeek -1
                constructWord word_to_return 0 file_in file_out

        return word_to_return

-- Obtiene una palabra del documento especificado,
-- copiando todos los caracteres no alfabeticos
-- precedentes al archivo de salida.
-- Cuando alcanza el final del documento, lo señaliza
-- con una excepcion.
doc_get_word :: Document -> IO Word
doc_get_word (Document file_in file_out) =

    catch (do
        valid_word <- constructWord "" file_in file_out
        return valid_word)
          (\err -> putStrLn(show err))


-- usar hgetchar no hace falta trabajar con la excepcion porq de eso se ocupa hgetchar
-- cuando llega a eof solo levanta la excepcion

-- Escribe una palabra en el documento de salida.
doc_put_word :: Word -> Document -> IO ()
doc_put_word word (Document file_in file_out) = 
    do
        hPutStr file_out word
        return()

File: lab5/spellchecker/src/document/Document.java
package document;
import word.Word;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;


/**
 * representa al documento que se va a procesar.
 */
public class Document {

	/** Atributes */
	private BufferedReader input_file = null;
	private BufferedWriter output_file = null;
	
	
	/** Constructor */
	public Document(String input_file, String output_file) throws IOException {
		this.input_file = new BufferedReader(new FileReader(input_file));
		this.output_file = new BufferedWriter(new FileWriter(output_file));
	}
	
	
	/** Methods */
	
	/**
	 * Cierra el documento.
	 */
	public void close() throws IOException {
		this.output_file.close();
		this.input_file.close();
	}
	
	/**
	 * El metodo getWord, lee el documento de entrada,
	 * palabra por palabra, copiando al documento de salida
	 * todos los caracteres no alfabeticos precedentes que encuentre.
	 * Al llegar al final del archivo, lo senaliza mediante una
	 * excepcion EOFException
	 */
	public Word getWord() throws IOException {
		String string_to_word= "";
		int c = 0;
		int i = 0;
		while((c = this.input_file.read()) != -1) {
			char character = (char) c;
			if (Character.isAlphabetic(character)) {
				string_to_word += character;
			}
			else {
				if (c != -1) {
					if (i == 0) {
						Word invalid_character = new Word(character+"");
						this.putWord(invalid_character);
					}
					else {
						System.out.println(character);
						this.input_file.mark(i-1);
						this.input_file.reset();
					}
					Word word_to_return = new Word(string_to_word);
					return word_to_return;
				}
			}
			i++;
		}
		// exit while
		return null;
	}
	
	/**
	 * Escribe la palabra w al archivo de salida.
	 */
	public void putWord(Word w) throws IOException {
		output_file.write(w.getWord());
	}
}
